esapi/esapi_util/esys_iutil.c

```diff
@@ -0,0 +1,1202 @@
+                  TSS2_ESYS_RC_BAD_TR);
+        return TSS2_ESYS_RC_BAD_TR;
+    }
+
+    /* There are special "global" object for the TPM, such as PCRs or
+       hierarchies. If they do not exist yet inside the Esys context we create
+       them here and return the newly created object. */
+    r = iesys_handle_to_tpm_handle(esys_handle, &tpm_handle);
+    return_if_error(r, "Unknown ESYS handle.");
+
+    r = esys_CreateResourceObject(esys_context, esys_handle, &esys_object_aux);
+    return_if_error(r, "Creating Resource Object.");
+
+    esys_object_aux->rsrc.handle = tpm_handle;
+    esys_object_aux->rsrc.rsrcType = IESYSC_WITHOUT_MISC_RSRC;
+
+    r = Tss2_MU_TPM2_HANDLE_Marshal(tpm_handle,
+                                &esys_object_aux->rsrc.name.name[0],
+                                sizeof(esys_object_aux->rsrc.name.name),
+                                &offset);
+    return_if_error(r, "Marshalling TPM handle.");
+
+    esys_object_aux->rsrc.name.size = offset;
+    *esys_object = esys_object_aux;
+    return TSS2_RC_SUCCESS;
+}
+
+/**
+ * Check that the esys context is ready for an _async call.
+ *
+ * This function will check that the sequence of invocations to the esys context
+ * was such that an _async function can be called. This means that the internal
+ * @state field is either @_ESYS_STATE_INIT, @_ESYS_STATE_ERRORRESPONSE,
+ * @_ESYS_STATE_FINISHED.
+ * @param[in,out] esys_context The esys context to issue the command on.
+ * @returns TSS2_RC_SUCCESS on success
+ *          TSS2_RC_BAD_SEQUENCE if context is not ready for this function
+ */
+TSS2_RC
+iesys_check_sequence_async(ESYS_CONTEXT * esys_context)
+{
+    if (esys_context == NULL) {
+        LOG_ERROR("esyscontext is NULL.");
+        return TSS2_ESYS_RC_BAD_REFERENCE;
+    }
+
+    if (esys_context->state != _ESYS_STATE_INIT &&
+        esys_context->state != _ESYS_STATE_ERRORRESPONSE &&
+        esys_context->state != _ESYS_STATE_FINISHED &&
+        esys_context->state != _ESYS_STATE_RESUBMISSION) {
+        LOG_ERROR("Esys called in bad sequence.");
+        return TSS2_ESYS_RC_BAD_SEQUENCE;
+    }
+//TODO: Check if RESUBMISSION BELONGS HERE OR RATHER INTO THE FINISH METHOD.
+    if (esys_context->state == _ESYS_STATE_RESUBMISSION) {
+        esys_context->submissionCount++;
+        LOG_DEBUG("The command will be resubmitted for the %i time.",
+                  esys_context->submissionCount);
+    } else {
+        esys_context->submissionCount = 0;
+    }
+    return TSS2_RC_SUCCESS;
+}
+
+TSS2_RC
+check_session_feasability(ESYS_TR shandle1, ESYS_TR shandle2, ESYS_TR shandle3,
+                          int mandatory)
+{
+    ESYS_TR handle_tab[3] = { shandle1, shandle2, shandle3 };
+    bool check_none = false;
+    for (int i = 2; i >= 0; i--) {
+        if (handle_tab[i] != ESYS_TR_NONE)
+            mandatory--;
+        if (handle_tab[i] != ESYS_TR_NONE || handle_tab[i] != ESYS_TR_PASSWORD)
```

Comment:
I think this is a tautology: `handle_tab[i]` cannot be both of these values at the same time.